### PR TITLE
Add border radius to UI nodes (adopted)

### PR DESCRIPTION
**Pull request by chompaa**
# Objective

Implements border radius for UI nodes. Adopted from #8973, but excludes shadows.

## Solution

- Add a component `BorderRadius` which contains a radius value for each corner of the UI node.
- Use a fragment shader to generate the rounded corners using a signed distance function.

</img>

## Changelog

- `BorderRadius`: New component that holds the border radius values.
- `NodeBundle` & `ButtonBundle`: Added a `border_radius: BorderRadius` field.
- `extract_uinode_borders`: Stripped down, most of the work is done in the shader now. Borders are no longer assembled from multiple rects, instead the shader uses a signed distance function to draw the border.
- `UiVertex`: Added size, border and radius fields.
- `UiPipeline`: Added three vertex attributes to the vertex buffer layout, to accept the UI node's size, border thickness and border radius.
- Examples: Added rounded corners to the UI element in the `button` example, and a `rounded_borders` example.

### CODE REVIEW

**Comment by viridia** [comment]
I think `UiBorderRadius` should be its own component. There has been a lot of confusion around the `Style` component, as to what exactly should and should not go in it, and part of the problem is that it's not well named. There's a proposal on the table to rename `Style` to `LayoutConstraints`: #10497 which should hopefully make it clearer - otherwise `Style` becomes a dumping ground for random properties that don't have an obvious home. All of the properties currently in `Style` are in fact layout constraints, but border radius is not a layout constraint.

**Comment by alice-i-cecile** [comment]
From previous conversation with @viridia about the precursor:

>  I looked at the rounded rect PR some more today, and have some further thoughts:
>
> 1. The changes to UiRect seem unnecessary, and are not required for rounded corners.
> 2. Part of what the PR is doing is changing how borders are rendered in general. Traditionally, borders have been rendered as a separate primitive. The PR instead changes it so that a single shader renders both borders and background, passing in separate colors for each. Assuming you want to go down that path, you could split out a separate PR to do just that part. However, you can also expect some push-back due to the extra per-vertex overhead needed.
> 3. I disagree with the author's decision to make border_radius part of Style, I think it should be a separate component as others have suggested.
> 4. As someone who uses rounded corners, my opinions on this PR are somewhat mixed. In my own work, about half of the use cases for rounded corners also require clipping, which this PR doesn't address (and which is really difficult to do without a full 2d clipping solution such as Vello provides).
> 5. I can get around the lack of clipping by writing my own custom UiMaterial for rounded corners, but if I'm going to do that, then I don't need rounded corner support in Bevy.

**Comment by alice-i-cecile** [comment]
I don't want to delay this unnecessarily, but here are my initial thoughts:

1. I really like your generated test cases in the PR description: can you add a variant of this as an example so users and maintainers can quickly check how this works?
2. I think that clipping out the background is required for this to be shippable: right now it looks fundamentally buggy.
3. Adjusting hit detection would be nice, but can be moved to future work.
4. Definitely agree on moving this out to its own component.

File: crates/bevy_ui/src/render/pipeline.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/pipeline.rs:68]: Add `border...` to each of these comments to make it more clear what they refer to.

File: crates/bevy_ui/src/render/render_pass.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/render_pass.rs:218]: Comments on this chunk of code would be helpful to explain what's going on.

File: crates/bevy_ui/src/ui_node.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/ui_node.rs:429]: The main docs belong on the `UiBorderRadius` type, not on this field.

File: crates/bevy_ui/src/ui_node.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/ui_node.rs:1873]: Missing docs.

File: crates/bevy_ui/src/ui_node.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/ui_node.rs:1875]:
```suggestion
    /// Zero curvature. All the corners will be right-angled.
```

File: crates/bevy_ui/src/ui_node.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/ui_node.rs:2069]: I don't like this: I think its value is limited and it's easy to screw up.

**Comment by viridia** [comment]
Assuming that you can get the rendering folks to sign off on the extra overhead involved in adding additional vertex params to UI node rendering, I would support this PR. It doesn't do everything I need, but a lot of people have wanted this for a long time, and as mentioned, I can always write custom shaders to fill the gaps.

What would be helpful, I think, is if we could post some actual perf data showing the impact. I'm not so familiar on how to do this, but I know that for my previous PR @cart posted some numbers.

**Comment by chompaa** [comment]
> I don't want to delay this unnecessarily, but here are my initial thoughts:

1. I've added an example `rounded_borders` analogous to `borders`, does this suffice?
2. Should be fixed now.
3. I'm not really sure how to approach this right now unfortunately.
4. I've moved `UiBorderRadius` to a component. Should this just be `BorderRadius` to be in line with `BorderColor` now?

File: crates/bevy_ui/src/render/render_pass.rs
Comment (chompaa) [reply crates/bevy_ui/src/render/render_pass.rs:218]: Would something like this be sufficient?

```rust
        // Store the vertices
        pass.set_vertex_buffer(0, ui_meta.vertices.buffer().unwrap().slice(..));
        // Define how to "connect" the vertices
        pass.set_index_buffer(
            ui_meta.indices.buffer().unwrap().slice(..),
            0,
            bevy_render::render_resource::IndexFormat::Uint32,
        );
        // Draw the vertices
        pass.draw_indexed(batch.range.clone(), 0, 0..1);
```

Or did you want something more comprehensive?

**Comment by viridia** [comment]
BTW, also wanted to mention that one of the issues that stalled out the previous PR was premultiplied alpha. Apparently the UI pipeline is configured in a way that requires colors to have premultiplied alpha. Normally this is invisible since alphas are almost always 1.0. However it can manifest as a light-colored "halo" around the rounded corners. The fix is simple: just multiply the computed RGB values by the alpha from the rounded corner algorithm. However, we should visually verify that this is needed before attempting to fix it.

**Comment by viridia** [comment]
@chompaa I think the `Ui` prefix is really only needed for cases where disambiguation is needed, e.g. `UiRect` vs `Rect`.

**Comment by chompaa** [comment]
> @chompaa I think the `Ui` prefix is really only needed for cases where disambiguation is needed, e.g. `UiRect` vs `Rect`.

Agreed, I'll rename shortly.

File: crates/bevy_ui/src/render/render_pass.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/render/render_pass.rs:218]: Yep that's great :)

Comment (chompaa) [reply crates/bevy_ui/src/render/render_pass.rs:218]: Great, these are in already since I accidentally committed them :)

File: examples/ui/rounded_borders.rs
Comment (alice-i-cecile) [inline examples/ui/rounded_borders.rs:87]: Can we swap this 64 out for a computed value to make this clearer and more robust to change?

File: crates/bevy_ui/src/render/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:147]: Doc strings here.

File: crates/bevy_ui/src/render/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:166]: Doc strings on these fields: the order of the floats is important and non-obvious.

File: crates/bevy_ui/src/render/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:787]: Doc strings here as well.

File: crates/bevy_ui/src/render/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:833]: These values should be moved to doc strings, rather than just comments :)

**Review by alice-i-cecile — APPROVED** [review]
Overall I'd be happy to see this merged in its current form. It has an example that we can use to validate that this works, the code quality is solid, and the critical functionality is there.

While I would like to see more aggressive docs (I know the surrounding code doesn't have it yet, but rendering docs really do need to improve), I'm okay if that doesn't happen in this PR.

If you're in the mood to shave yaks, we can add text labels to all of the UI border radius nodes writing out what they should look like.

File: examples/ui/rounded_borders.rs
Comment (chompaa) [reply examples/ui/rounded_borders.rs:87]: Agreed, I didn't change this since this was how the `borders` example was done. Since we have all border combinations defined above, should this loop just be:

```rust
// let borders = ...
for border in borders {
  // ...
}
```

**Comment by chompaa** [comment]
> If you're in the mood to shave yaks, we can add text labels to all of the UI border radius nodes writing out what they should look like.

I'm happy to go over what things should look like :)

File: crates/bevy_ui/src/render/mod.rs
Comment (chompaa) [reply crates/bevy_ui/src/render/mod.rs:833]: I suppose I should also remove the `BOX_SHADOW` constant here, since it's just an artifact from the original PR anyway.

File: crates/bevy_ui/src/render/mod.rs
Comment (chompaa) [reply crates/bevy_ui/src/render/mod.rs:166]: What about:
```rust
/// Border radius of the UI node.
/// Ordering: top left, top right, bottom right, bottom left.
pub border_radius: [f32; 4],
```

File: examples/ui/rounded_borders.rs
Comment (alice-i-cecile) [reply examples/ui/rounded_borders.rs:87]: Yeah that would be perfect.

File: crates/bevy_ui/src/render/mod.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/render/mod.rs:166]: That's exactly what I was hoping for.

**Comment by alice-i-cecile** [comment]
> > If you're in the mood to shave yaks, we can add text labels to all of the UI border radius nodes writing out what they should look like.
> 
> I'm happy to go over what things should look like :)

Awesome. I think just a small text label under each square that says what the config is would work great. We can make a follow-up issue to clean up the `borders` example to match once this is merged :)

**Comment by alice-i-cecile** [comment]
Looks great. If you're up for it,  grab a new screenshot and add it to the PR description: we'll use that to populate the blog post.

**Comment by chompaa** [comment]
> Looks great. If you're up for it, grab a new screenshot and add it to the PR description: we'll use that to populate the blog post.

Alright, will do shortly.

**Review by JMS55 — APPROVED** [review]
Lgtm. I didn't compare the shader code to the references thoroughly, but so long as the demo works I'm fine with it. 

I would like to put links to the references in the shader code though.

File: crates/bevy_ui/src/ui_node.rs
Comment (bushrat011899) [inline crates/bevy_ui/src/ui_node.rs:2065]:
```suggestion
    pub const fn new(top_left: Val, top_right: Val, bottom_right: Val, bottom_left: Val) -> Self {
        Self {
            top_left,
            top_right,
            bottom_right,
            bottom_left,
        }
    }

    #[inline]
    /// Sets the radii to logical pixel values.
    pub const fn px(top_left: f32, top_right: f32, bottom_right: f32, bottom_left: f32) -> Self {
        Self {
            top_left: Val::Px(top_left),
            top_right: Val::Px(top_right),
            bottom_right: Val::Px(bottom_right),
            bottom_left: Val::Px(bottom_left),
        }
    }

    #[inline]
    /// Sets the radii to percentage values.
    pub const fn percent(top_left: f32, top_right: f32, bottom_right: f32, bottom_left: f32) -> Self {
        Self {
            top_left: Val::Px(top_left),
            top_right: Val::Px(top_right),
            bottom_right: Val::Px(bottom_right),
            bottom_left: Val::Px(bottom_left),
        }
    }

    #[inline]
    /// Sets the radius for the top left corner.
    /// Remaining corners will be right-angled.
    pub const fn top_left(radius: Val) -> Self {
        Self {
            top_left: radius,
            ..Self::DEFAULT
        }
    }

    #[inline]
    /// Sets the radius for the top right corner.
    /// Remaining corners will be right-angled.
    pub const fn top_right(radius: Val) -> Self {
        Self {
            top_right: radius,
            ..Self::DEFAULT
        }
    }

    #[inline]
    /// Sets the radius for the bottom right corner.
    /// Remaining corners will be right-angled.
    pub const fn bottom_right(radius: Val) -> Self {
        Self {
            bottom_right: radius,
            ..Self::DEFAULT
        }
    }

    #[inline]
    /// Sets the radius for the bottom left corner.
    /// Remaining corners will be right-angled.
    pub const fn bottom_left(radius: Val) -> Self {
        Self {
            bottom_left: radius,
            ..Self::DEFAULT
        }
    }

    #[inline]
    /// Sets the radii for the top left and bottom left corners.
    /// Remaining corners will be right-angled.
    pub const fn left(radius: Val) -> Self {
        Self {
            top_left: radius,
            bottom_left: radius,
            ..Self::DEFAULT
        }
    }

    #[inline]
    /// Sets the radii for the top right and bottom right corners.
    /// Remaining corners will be right-angled.
    pub const fn right(radius: Val) -> Self {
        Self {
            top_right: radius,
            bottom_right: radius,
            ..Self::DEFAULT
        }
    }

    #[inline]
    /// Sets the radii for the top left and top right corners.
    /// Remaining corners will be right-angled.
    pub const fn top(radius: Val) -> Self {
        Self {
            top_left: radius,
            top_right: radius,
            ..Self::DEFAULT
        }
    }

    #[inline]
    /// Sets the radii for the bottom left and bottom right corners.
    /// Remaining corners will be right-angled.
    pub const fn bottom(radius: Val) -> Self {
        Self {
            bottom_left: radius,
            bottom_right: radius,
            ..Self::DEFAULT
        }
    }

    /// Returns the [`BorderRadius`] with its `top_left` field set to the given value.
    #[inline]
    pub const fn with_top_left(mut self, radius: Val) -> Self {
        self.top_left = radius;
        self
    }

    /// Returns the [`BorderRadius`] with its `top_right` field set to the given value.
    #[inline]
    pub const fn with_top_right(mut self, radius: Val) -> Self {
        self.top_right = radius;
        self
    }

    /// Returns the [`BorderRadius`] with its `bottom_right` field set to the given value.
    #[inline]
    pub const fn with_bottom_right(mut self, radius: Val) -> Self {
        self.bottom_right = radius;
        self
    }

    /// Returns the [`BorderRadius`] with its `bottom_left` field set to the given value.
    #[inline]
    pub const fn with_bottom_left(mut self, radius: Val) -> Self {
        self.bottom_left = radius;
        self
    }

    /// Returns the [`BorderRadius`] with its `top_left` and `bottom_left` fields set to the given value.
    #[inline]
    pub const fn with_left(mut self, radius: Val) -> Self {
        self.top_left = radius;
        self.bottom_left = radius;
        self
    }

    /// Returns the [`BorderRadius`] with its `top_right` and `bottom_right` fields set to the given value.
    #[inline]
    pub const fn with_right(mut self, radius: Val) -> Self {
        self.top_right = radius;
        self.bottom_right = radius;
        self
    }

    /// Returns the [`BorderRadius`] with its `top_left` and `top_right` fields set to the given value.
    #[inline]
    pub const fn with_top(mut self, radius: Val) -> Self {
        self.top_left = radius;
        self.top_right = radius;
        self
    }

    /// Returns the [`BorderRadius`] with its `bottom_left` and `bottom_right` fields set to the given value.
    #[inline]
    pub const fn with_bottom(mut self, radius: Val) -> Self {
```

Sorry for the ugly suggestion! All these methods can be `const` so they probably should be.

**Review by bushrat011899 — COMMENTED** [review]
I'm not familiar enough with the UI system to feel confident approving this PR, but I didn't notice anything particularly wrong. Only suggestion is to make the `BorderRadius` methods all `const`, since they all can be, and likely won't ever need non-`const` features in the future.

File: crates/bevy_ui/src/ui_node.rs
Comment (pablo-lua) [inline crates/bevy_ui/src/ui_node.rs:1856]:
```suggestion
#[reflect(PartialEq, Default)]
```

**Review by pablo-lua — COMMENTED** [review]
Just a reflection comment.

File: crates/bevy_ui/src/render/mod.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/render/mod.rs:831]: Yep, that works well.

**Comment by alice-i-cecile** [comment]
Awesome, merging now: we can refine this more before 0.14 if issues come up <3

Thank you, I've wanted this feature for literally years.

**Pull request merged**
